app/lib/common/toolkit/html_editor/mentions/mention_menu.dart

```diff
@@ -0,0 +1,299 @@
+import 'dart:async';
+import 'dart:math';
+
+import 'package:acter/common/toolkit/html_editor/mentions/widgets/mention_list.dart';
+import 'package:acter/common/toolkit/html_editor/mentions/models/mention_type.dart';
+import 'package:acter/common/toolkit/html_editor/mentions/selected_mention_provider.dart';
+import 'package:acter/common/toolkit/html_editor/services/constants.dart';
+import 'package:appflowy_editor/appflowy_editor.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+MentionMenu? _menu;
+
+class MentionMenu {
+  MentionMenu({
+    required this.editorState,
+    required this.roomId,
+    required this.mentionType,
+    required this.ref,
+  });
+
+  final EditorState editorState;
+  final String roomId;
+  final MentionType mentionType;
+  final WidgetRef ref;
+  OverlayEntry? _menuEntry;
+  StreamSubscription<EditorTransactionValue>? _updateListener;
+
+  bool selectionChangedByMenu = false;
+
+  static bool isShowing() => _menu != null;
+
+  static void dismiss() {
+    _menu?.hide();
+    _menu = null;
+  }
+
+  void show(BuildContext context) {
+    if (_menu != null) return;
+    _listenToEditor(context);
+    _show(context);
+  }
+
+  void hide() {
+    // clearing all the things
+    _updateListener?.cancel();
+    _updateListener = null;
+    _menuEntry?.remove();
+    _menuEntry = null;
+  }
+
+  // control functions
+  static void next() {
+    _menu?.provider.next();
+  }
+
+  static void prev() {
+    _menu?.provider.prev();
+  }
+
+  static void selectCurrent() {
+    if (_menu?._selectCurrent() ?? false) {
+      return;
+    }
+    // selectin failed, just hide
+    _menu?.hide();
+  }
+
+  bool _selectCurrent() {
+    final filtered = switch (mentionType) {
+      MentionType.user => ref.read(filteredUserSuggestionsProvider(roomId)),
+      MentionType.room => ref.read(filteredRoomSuggestionsProvider(roomId)),
+    };
+    final selected = max(
+      0, // 0 or higher
+      switch (mentionType) {
+            MentionType.user => ref.read(selecteUserMentionProvider(roomId)),
+            MentionType.room => ref.read(selectedRoomMentionProvider(roomId)),
+          } ??
+          0, // select the first item as fallback
+    );
+    if (selected >= filtered.length) return false; // too long, ignore
+    final item = filtered.keys.elementAt(selected);
+    _select(mentionType, item, filtered[item]);
+    return true;
+  }
+
+  SelectedMentionNotifier get provider => switch (mentionType) {
+    MentionType.user => ref.read(selecteUserMentionProvider(roomId).notifier),
+    MentionType.room => ref.read(selectedRoomMentionProvider(roomId).notifier),
+  };
```

Comment:
this isn't great IMHO but the fastest way within the existing infrastructure. I see us replacing this with a different mechanism in a future PR when we want to add object and room mention support...